File: surface/widgets/appkit/components/widgets/concepts/process-human/ApprovalStepper.swift
import AppKit

class ApprovalStepperView: NSView {

    enum State: String { case viewing; case stepFocused; case acting }

    struct Step {
        let id: String
        let label: String
        let approver: String?
        let status: String   // pending | approved | rejected | skipped | active
        let timestamp: String?
        let quorumRequired: Int?
        let quorumCurrent: Int?
    }

    private(set) var state: State = .viewing
    private var steps: [Step] = []
    private var currentStepId: String? = nil
    private var focusedStepId: String? = nil
    private var actingStepId: String? = nil
    private var focusIndex: Int = 0
    private var dueAt: Date? = nil
    private var slaTimer: Timer?

    var onApprove: ((String) -> Void)?
    var onReject: ((String) -> Void)?
    var onDelegate: ((String) -> Void)?
    var onClaim: ((String) -> Void)?

    private let rootStack = NSStackView()
    private let stepScroll = NSScrollView()
    private let stepContainer = NSStackView()
    private let actionBar = NSStackView()
    private let approveBtn = NSButton(title: "Approve", target: nil, action: nil)
    private let rejectBtn = NSButton(title: "Reject", target: nil, action: nil)
    private let delegateBtn = NSButton(title: "Delegate", target: nil, action: nil)
    private let cancelBtn = NSButton(title: "Cancel", target: nil, action: nil)
    private let slaLabel = NSTextField(labelWithString: "")

    func reduce(_ event: String, stepId: String? = nil) {
        switch state {
        case .viewing:
            if event == "FOCUS_STEP", let id = stepId { state = .stepFocused; focusedStepId = id }
            if event == "START_ACTION" { state = .acting; actingStepId = stepId ?? focusedStepId }
        case .stepFocused:
            if event == "BLUR" { state = .viewing; focusedStepId = nil }
            if event == "START_ACTION" { state = .acting; actingStepId = stepId ?? focusedStepId }
            if event == "FOCUS_STEP", let id = stepId { focusedStepId = id }
        case .acting:
            if event == "COMPLETE" { state = .viewing; actingStepId = nil }
            if event == "CANCEL" { state = .viewing; actingStepId = nil }
        }
        updateUI()
    }

    override init(frame: NSRect) {
        super.init(frame: frame); setupSubviews()
        setAccessibilityRole(.list); setAccessibilityLabel("Multi-step approval flow visualization")
    }
    required init?(coder: NSCoder) { super.init(coder: coder) }

    private func setupSubviews() {
        wantsLayer = true
        rootStack.orientation = .vertical; rootStack.spacing = 6
        rootStack.translatesAutoresizingMaskIntoConstraints = false
        addSubview(rootStack)
        NSLayoutConstraint.activate([
            rootStack.topAnchor.constraint(equalTo: topAnchor, constant: 8),
            rootStack.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 8),
            rootStack.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -8),
            rootStack.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -8),
        ])

        stepScroll.hasVerticalScroller = true; stepScroll.drawsBackground = false
        stepContainer.orientation = .vertical; stepContainer.spacing = 4
        stepScroll.documentView = stepContainer
        rootStack.addArrangedSubview(stepScroll)

        // Action bar
        actionBar.orientation = .horizontal; actionBar.spacing = 6
        approveBtn.bezelStyle = .roundRect; approveBtn.target = self; approveBtn.action = #selector(handleApprove(_:))
        rejectBtn.bezelStyle = .roundRect; rejectBtn.target = self; rejectBtn.action = #selector(handleReject(_:))
        delegateBtn.bezelStyle = .roundRect; delegateBtn.target = self; delegateBtn.action = #selector(handleDelegate(_:))
        cancelBtn.bezelStyle = .roundRect; cancelBtn.target = self; cancelBtn.action = #selector(handleCancel(_:))
        actionBar.addArrangedSubview(approveBtn); actionBar.addArrangedSubview(rejectBtn)
        actionBar.addArrangedSubview(delegateBtn); actionBar.addArrangedSubview(cancelBtn)
        actionBar.isHidden = true
        rootStack.addArrangedSubview(actionBar)

        // SLA indicator
        slaLabel.font = .systemFont(ofSize: 11); slaLabel.textColor = .secondaryLabelColor
        slaLabel.isHidden = true
        rootStack.addArrangedSubview(slaLabel)

        updateUI()
    }

    func configure(steps: [Step], currentStep: String, dueAt: Date? = nil) {
        self.steps = steps; self.currentStepId = currentStep; self.dueAt = dueAt
        rebuildSteps(); startSLATimer(); updateUI()
    }

    private func stepStatusIcon(_ status: String) -> String {
        switch status {
        case "approved": return "\u{2713}"
        case "rejected": return "\u{2717}"
        case "skipped": return "\u{2014}"
        case "active": return "\u{25CF}"
        default: return "\u{25CB}"
        }
    }

    private func statusColor(_ status: String) -> NSColor {
        switch status {
        case "approved": return .systemGreen
        case "rejected": return .systemRed
        case "active": return .systemBlue
        case "skipped": return .secondaryLabelColor
        default: return .tertiaryLabelColor
        }
    }

    private func rebuildSteps() {
        stepContainer.arrangedSubviews.forEach { $0.removeFromSuperview() }
        for (i, step) in steps.enumerated() {
            let isCurrent = step.id == currentStepId
            let isFocused = focusIndex == i

            let row = NSStackView(); row.orientation = .horizontal; row.spacing = 8
            row.wantsLayer = true
            if isFocused || isCurrent { row.layer?.backgroundColor = NSColor.controlAccentColor.withAlphaComponent(0.08).cgColor }

            // Indicator
            let indicator = NSTextField(labelWithString: (step.status == "pending" || step.status == "active") ? "\(i + 1)" : stepStatusIcon(step.status))
            indicator.font = .boldSystemFont(ofSize: 13); indicator.textColor = statusColor(step.status)
            indicator.widthAnchor.constraint(equalToConstant: 24).isActive = true; indicator.alignment = .center
            row.addArrangedSubview(indicator)

            // Label button
            let nameBtn = NSButton(title: step.label, target: self, action: #selector(handleStepClick(_:)))
            nameBtn.bezelStyle = .roundRect; nameBtn.tag = i
            nameBtn.setAccessibilityLabel("Step \(i + 1): \(step.label) \u{2014} \(step.status)")
            row.addArrangedSubview(nameBtn)

            // Status badge
            let statusLabel = NSTextField(labelWithString: step.status.capitalized)
            statusLabel.font = .systemFont(ofSize: 10); statusLabel.textColor = statusColor(step.status)
            statusLabel.widthAnchor.constraint(equalToConstant: 60).isActive = true
            row.addArrangedSubview(statusLabel)

            // Approver
            if let approver = step.approver {
                let approverLabel = NSTextField(labelWithString: approver)
                approverLabel.font = .systemFont(ofSize: 10); approverLabel.textColor = .secondaryLabelColor
                row.addArrangedSubview(approverLabel)
            }

            // Quorum
            if let required = step.quorumRequired {
                let current = step.quorumCurrent ?? 0
                let quorumLabel = NSTextField(labelWithString: "\(current)/\(required)")
                quorumLabel.font = .monospacedDigitSystemFont(ofSize: 10, weight: .regular)
                quorumLabel.textColor = .secondaryLabelColor
                quorumLabel.setAccessibilityLabel("\(current) of \(required) approvals")
                row.addArrangedSubview(quorumLabel)
            }

            // Timestamp
            if let ts = step.timestamp {
                let tsLabel = NSTextField(labelWithString: ts)
                tsLabel.font = .systemFont(ofSize: 10); tsLabel.textColor = .tertiaryLabelColor
                row.addArrangedSubview(tsLabel)
            }

            stepContainer.addArrangedSubview(row)

            // Connector line between steps
            if i < steps.count - 1 {
                let connector = NSTextField(labelWithString: "\u{2502}")
                connector.font = .systemFont(ofSize: 10); connector.alignment = .center
                connector.textColor = statusColor(step.status)
                stepContainer.addArrangedSubview(connector)
            }
        }
    }

    private func startSLATimer() {
        slaTimer?.invalidate(); slaTimer = nil
        guard dueAt != nil else { slaLabel.isHidden = true; return }
        slaLabel.isHidden = false
        updateSLA()
        slaTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in self?.updateSLA() }
    }

    private func updateSLA() {
        guard let due = dueAt else { return }
        let diff = due.timeIntervalSinceNow
        if diff <= 0 { slaLabel.stringValue = "SLA: Overdue"; slaLabel.textColor = .systemRed; return }
        let hours = Int(diff) / 3600; let minutes = (Int(diff) % 3600) / 60
        if hours > 24 { slaLabel.stringValue = "SLA: \(hours / 24)d \(hours % 24)h"; slaLabel.textColor = .secondaryLabelColor }
        else if hours > 0 { slaLabel.stringValue = "SLA: \(hours)h \(minutes)m"; slaLabel.textColor = hours < 4 ? .systemOrange : .secondaryLabelColor }
        else { slaLabel.stringValue = "SLA: \(minutes)m"; slaLabel.textColor = .systemRed }
    }

    private func updateUI() {
        actionBar.isHidden = state != .acting
        rebuildSteps()
        setAccessibilityValue("\(steps.count) steps, \(state.rawValue)")
    }

    @objc private func handleStepClick(_ sender: NSButton) {
        let idx = sender.tag; guard idx < steps.count else { return }
        focusIndex = idx
        reduce("FOCUS_STEP", stepId: steps[idx].id)
    }
    @objc private func handleApprove(_ sender: NSButton) {
        if let id = actingStepId { onApprove?(id) }; reduce("COMPLETE")
    }
    @objc private func handleReject(_ sender: NSButton) {
        if let id = actingStepId { onReject?(id) }; reduce("COMPLETE")
    }
    @objc private func handleDelegate(_ sender: NSButton) {
        if let id = actingStepId { onDelegate?(id) }; reduce("COMPLETE")
    }
    @objc private func handleCancel(_ sender: NSButton) { reduce("CANCEL") }

    override var acceptsFirstResponder: Bool { true }
    override func keyDown(with event: NSEvent) {
        switch event.keyCode {
        case 125: // down
            focusIndex = min(focusIndex + 1, steps.count - 1)
            if focusIndex < steps.count { reduce("FOCUS_STEP", stepId: steps[focusIndex].id) }
        case 126: // up
            focusIndex = max(focusIndex - 1, 0)
            if focusIndex < steps.count { reduce("FOCUS_STEP", stepId: steps[focusIndex].id) }
        case 36: // enter
            let step = steps[focusIndex]
            if step.status == "active" || step.id == currentStepId { reduce("START_ACTION", stepId: step.id) }
        case 53: // escape
            if state == .acting { reduce("CANCEL") } else { reduce("BLUR") }
        default: super.keyDown(with: event)
        }
    }
    deinit { slaTimer?.invalidate() }
}

File: surface/widgets/vue/components/widgets/concepts/process-human/ApprovalStepper.ts
import { defineComponent, h, ref, computed, type PropType } from 'vue';

export type ApprovalStepperState = 'viewing' | 'stepFocused' | 'acting';
export type ApprovalStepperEvent =
  | { type: 'FOCUS_STEP'; id?: string }
  | { type: 'START_ACTION' }
  | { type: 'BLUR' }
  | { type: 'COMPLETE' }
  | { type: 'CANCEL' };

export function approvalStepperReducer(state: ApprovalStepperState, event: ApprovalStepperEvent): ApprovalStepperState {
  switch (state) {
    case 'viewing':
      if (event.type === 'FOCUS_STEP') return 'stepFocused';
      if (event.type === 'START_ACTION') return 'acting';
      return state;
    case 'stepFocused':
      if (event.type === 'BLUR') return 'viewing';
      if (event.type === 'START_ACTION') return 'acting';
      return state;
    case 'acting':
      if (event.type === 'COMPLETE') return 'viewing';
      if (event.type === 'CANCEL') return 'viewing';
      return state;
    default:
      return state;
  }
}

interface ApprovalStep {
  id: string;
  label: string;
  approver?: string;
  status: 'pending' | 'approved' | 'rejected' | 'skipped' | 'active';
  timestamp?: string;
  quorumRequired?: number;
  quorumCurrent?: number;
}

function stepStatusIcon(status: string): string {
  switch (status) { case 'approved': return '\u2713'; case 'rejected': return '\u2717'; case 'skipped': return '\u2014'; case 'active': return '\u25CF'; default: return '\u25CB'; }
}

function connectorStatus(prevStatus: string): string {
  switch (prevStatus) { case 'approved': return 'complete'; case 'rejected': return 'rejected'; case 'active': return 'active'; default: return 'pending'; }
}

function formatTimeRemaining(dueAt: string): string {
  const diff = new Date(dueAt).getTime() - Date.now();
  if (diff <= 0) return 'Overdue';
  const hours = Math.floor(diff / 3600000);
  const minutes = Math.floor((diff % 3600000) / 60000);
  if (hours > 24) return `${Math.floor(hours / 24)}d ${hours % 24}h`;
  if (hours > 0) return `${hours}h ${minutes}m`;
  return `${minutes}m`;
}

export const ApprovalStepper = defineComponent({
  name: 'ApprovalStepper',
  props: {
    steps: { type: Array as PropType<ApprovalStep[]>, required: true },
    currentStep: { type: String, required: true },
    status: { type: String, required: true },
    assignee: { type: String, default: undefined },
    dueAt: { type: String, default: undefined },
    variant: { type: String as PropType<'sequential' | 'parallel' | 'mixed'>, default: 'sequential' },
    orientation: { type: String as PropType<'horizontal' | 'vertical'>, default: 'horizontal' },
    showSLA: { type: Boolean, default: true },
    showAssignee: { type: Boolean, default: true },
  },
  emits: ['approve', 'reject', 'delegate', 'claim'],
  setup(props, { emit, slots }) {
    const state = ref<ApprovalStepperState>('viewing');
    const focusIndex = ref(0);
    const actingStepId = ref<string | null>(null);

    function send(event: ApprovalStepperEvent) {
      state.value = approvalStepperReducer(state.value, event);
    }

    function handleFocusStep(id: string, index: number) {
      focusIndex.value = index;
      send({ type: 'FOCUS_STEP', id });
    }

    function handleStartAction(stepId: string) {
      actingStepId.value = stepId;
      send({ type: 'START_ACTION' });
    }

    function handleApprove(stepId: string) {
      emit('approve', stepId);
      send({ type: 'COMPLETE' });
      actingStepId.value = null;
    }

    function handleReject(stepId: string) {
      emit('reject', stepId);
      send({ type: 'COMPLETE' });
      actingStepId.value = null;
    }

    function handleDelegate(stepId: string) {
      emit('delegate', stepId);
      send({ type: 'COMPLETE' });
      actingStepId.value = null;
    }

    function handleCancelAction() {
      send({ type: 'CANCEL' });
      actingStepId.value = null;
    }

    function handleKeydown(e: KeyboardEvent) {
      const nextKey = props.orientation === 'horizontal' ? 'ArrowRight' : 'ArrowDown';
      const prevKey = props.orientation === 'horizontal' ? 'ArrowLeft' : 'ArrowUp';
      if (e.key === nextKey || e.key === 'ArrowDown') { e.preventDefault(); const ni = Math.min(focusIndex.value + 1, props.steps.length - 1); handleFocusStep(props.steps[ni].id, ni); }
      if (e.key === prevKey || e.key === 'ArrowUp') { e.preventDefault(); const ni = Math.max(focusIndex.value - 1, 0); handleFocusStep(props.steps[ni].id, ni); }
      if (e.key === 'Enter') { e.preventDefault(); const step = props.steps[focusIndex.value]; if (step && (step.status === 'active' || step.id === props.currentStep)) handleStartAction(step.id); }
      if (e.key === 'Escape') { e.preventDefault(); handleCancelAction(); }
    }

    return () => {
      const children: any[] = [];

      // Step list
      const stepNodes = props.steps.map((step, index) => {
        const isCurrent = step.id === props.currentStep;
        const isFocused = focusIndex.value === index;
        const isActing = actingStepId.value === step.id && state.value === 'acting';
        const showConnector = index < props.steps.length - 1;

        const stepChildren: any[] = [];

        // Step node
        const nodeChildren: any[] = [
          h('div', { 'data-part': 'step-indicator', 'data-index': index + 1, 'data-status': step.status, 'aria-hidden': 'true' }, [
            h('span', {}, step.status === 'pending' || step.status === 'active' ? String(index + 1) : stepStatusIcon(step.status)),
          ]),
          h('span', { 'data-part': 'step-label' }, step.label),
        ];

        if (props.showAssignee && step.approver) {
          nodeChildren.push(h('div', { 'data-part': 'step-assignee', 'data-visible': 'true' }, [h('span', { 'data-part': 'assignee-name' }, step.approver)]));
        }

        nodeChildren.push(h('div', { 'data-part': 'step-status', 'data-status': step.status }, [h('span', {}, step.status)]));

        if (props.variant !== 'sequential' && step.quorumRequired) {
          nodeChildren.push(h('div', { 'data-part': 'quorum-display', 'aria-label': `${step.quorumCurrent ?? 0} of ${step.quorumRequired} approvals` }, [
            h('span', { 'data-part': 'quorum-count' }, `${step.quorumCurrent ?? 0}/${step.quorumRequired}`),
          ]));
        }

        if (step.timestamp) nodeChildren.push(h('span', { 'data-part': 'step-timestamp' }, step.timestamp));

        stepChildren.push(h('div', {
          'data-part': 'step', role: 'listitem',
          'aria-current': isCurrent ? 'step' : undefined,
          'aria-label': `Step ${index + 1}: ${step.label} \u2014 ${step.status}`,
          'data-status': step.status, 'data-current': isCurrent ? 'true' : 'false',
          tabindex: isFocused ? 0 : -1,
          onClick: () => handleFocusStep(step.id, index),
          onDblclick: () => { if (isCurrent || step.status === 'active') handleStartAction(step.id); },
        }, nodeChildren));

        // Action bar
        if (isActing) {
          stepChildren.push(h('div', { 'data-part': 'action-bar', role: 'toolbar', 'aria-label': 'Approval actions', 'data-visible': 'true' }, [
            h('button', { type: 'button', 'data-part': 'approve-button', onClick: () => handleApprove(step.id), 'aria-label': `Approve step: ${step.label}` }, 'Approve'),
            h('button', { type: 'button', 'data-part': 'reject-button', onClick: () => handleReject(step.id), 'aria-label': `Reject step: ${step.label}` }, 'Reject'),
            h('button', { type: 'button', 'data-part': 'delegate-button', onClick: () => handleDelegate(step.id), 'aria-label': `Delegate step: ${step.label}` }, 'Delegate'),
            h('button', { type: 'button', 'data-part': 'cancel-action-button', onClick: handleCancelAction, 'aria-label': 'Cancel' }, 'Cancel'),
          ]));
        }

        // Connector
        if (showConnector) {
          stepChildren.push(h('div', { 'data-part': 'connector', 'data-orientation': props.orientation, 'data-status': connectorStatus(step.status), 'aria-hidden': 'true' }));
        }

        return h('div', { key: step.id, 'data-part': 'step-wrapper' }, stepChildren);
      });

      children.push(h('div', { 'data-part': 'step-list', 'data-orientation': props.orientation }, stepNodes));

      // SLA
      if (props.showSLA && props.dueAt) {
        const overdue = new Date(props.dueAt).getTime() < Date.now();
        children.push(h('div', {
          'data-part': 'sla-indicator', role: 'timer', 'aria-label': 'Time remaining',
          'data-visible': 'true', 'data-due': props.dueAt, 'data-overdue': overdue ? 'true' : 'false',
        }, [
          h('span', { 'data-part': 'sla-label' }, 'SLA: '),
          h('span', { 'data-part': 'sla-countdown' }, formatTimeRemaining(props.dueAt)),
        ]));
      }

      if (slots.default) children.push(slots.default());

      return h('div', {
        role: 'list',
        'aria-label': 'Approval steps',
        'data-surface-widget': '',
        'data-widget-name': 'approval-stepper',
        'data-part': 'root',
        'data-variant': props.variant,
        'data-orientation': props.orientation,
        'data-state': state.value,
        tabindex: 0,
        onKeydown: handleKeydown,
      }, children);
    };
  },
});

export default ApprovalStepper;
